third_party/ienn/src/utils.cpp

```diff
@@ -0,0 +1,369 @@
+
+  // According to Android NN API doc:
+  // If rank is greater than 2, then it gets flattened to a 2-D Tensor.
+  // The (flattened) 2-D Tensor is reshaped (if necessary) to
+  // [batch_size, input_size], where "input_size" corresponds to the number of
+  // inputs to the layer, matching the second dimension of weights, and
+  // "batch_size" is calculated by dividing the number of elements by
+  // "input_size".
+  if (input.dimensions.size() > 2) {
+    params.input_batch_size = product(input.dimensions) / params.input_size;
+  } else {
+    if (input.dimensions[1] != params.input_size) {
+      std::cout << "input.dimensions[1] (" << input.dimensions[1] << ") "
+                << "!= input_size (" << params.input_size << ")";
+      return BAD_DATA;
+    }
+    params.input_batch_size = input.dimensions[0];
+  }
+
+  // A 1-D tensor, of shape [num_units]
+  const uint32_t bias_idx = inputs[index++];
+  const Operand& bias = model->operands[bias_idx];
+  params.bias_num_units = bias.dimensions[0];
+  if (params.bias_num_units != params.num_units) {
+    std::cout << "bias_num_units (" << params.bias_num_units << ") != "
+              << "num_units (" << params.num_units << ")";
+    return BAD_DATA;
+  }
+
+  params.fuse_code = GetScalarInt32(model, inputs[index++]);
+
+  return NOT_ERROR;
+}
+
+int32_t GetResizeBilinearParams(ModelInfoPtr model,
+                                const Operation& operation,
+                                ResizeBilinearParams& params) {
+  const int32_t type = operation.type;
+  if (operation.type != operation_t::RESIZE_BILINEAR_NN) {
+    std::cout << "Operation type " << type << " is not resize bilinear";
+    return BAD_DATA;
+  }
+  const std::vector<uint32_t>& inputs = operation.inputs;
+  if (inputs.size() != 3 && inputs.size() != 4) {
+    std::cout << "Inputs size is wrong " << inputs.size();
+    return BAD_DATA;
+  }
+  const Operand& input = model->operands[inputs[0]];
+  if (input.dimensions.size() != 4) {
+    std::cout << "Input must be a 4-D tensor";
+    return BAD_DATA;
+  }
+  params.height = input.dimensions[1];
+  params.width = input.dimensions[2];
+  params.new_height = GetScalarInt32(model, inputs[1]);
+  params.new_width = GetScalarInt32(model, inputs[2]);
+  params.y_scale = params.new_height / params.height;
+  params.x_scale = params.new_width / params.width;
+
+  params.align_corners = false;
+  if (inputs.size() == 4) {
+    params.align_corners = GetScalarInt32(model, inputs[3]) == 0 ? false : true;
+  }
+
+  return NOT_ERROR;
+}
+
+int32_t GetArgmaxParams(ModelInfoPtr model,
+                        const Operation& operation,
+                        ArgmaxParams& params) {
+  const int32_t type = operation.type;
+  if (type != ARGMAX) {
+    std::cout << "Operation type " << type << " is not argmax";
+    return BAD_DATA;
+  }
+  const std::vector<uint32_t>& inputs = operation.inputs;
+  params.axis = GetScalarInt32(model, inputs[1]);
+  return NOT_ERROR;
+}
```

Comment:
Does this file duplicate many implementations of https://github.com/otcshare/chromium-src/blob/webml/services/ml/compilation_impl.h?

Reply:
Yes, but there are different function declaration, Maybe we can refactor later to reuse those code.

int32_t GetArgmaxParams(const **mojom::OperationPtr**&, ArgmaxParams&) const;

int32_t GetArgmaxParams(**ModelInfoPtr model, const Operation**& operation, ArgmaxParams& params)

Reply:
How about other functions? I am fine to refactor them later.